design/external-references.md

```diff
@@ -0,0 +1,394 @@
+}
+```
+
+The deserialization context is an object that implements the following interface:
+
+```ts
+interface IDeserializationContext {
+  scope: Construct;
+  id: string;
+  readString(key: string, options?: ReadStringOptions): string;
+  readObject(key: string): IDeserializationContext;
+}
+
+interface ReadStringOptions {
+  allowUnresolved?: boolean; /** @default true */
+}
+```
+
+`readString(key)` can be used to read values stored by `writeString`. and `readObject(key)` returns a deserialization context for composite deserialization written via `writeObject`.
+
+The `allowUnresolved` option for `readString` can be used by constructs to indicate that returned value must be a resolved value (i.e. not a token). This implies, for example, that when importing this value, users cannot use the `resolveType: Deployment`  option (REQ8).
+
+Since `deserializeXxx` will need to create new construct objects, the deserialization context will supply a consistent `scope` and `id` which can be used to instantiate a construct object that represents this object. For example, `scope` can be mapped to the current `Stack` and `id` can be mapped to `exportName` which is ensured to be unique within the environment (and therefore, the current stack).
+
+Implementers of `deserializeXxx` should check if a construct with `id` already exists within `scope` and return it instead of instantiating and new object (REQ6).
+
+---
+
+Here's an example that demonstrates how serialization and deserialization can be implemented for an application load balancer (ALB), which is a composite resource. As you can see, ALBs encapsulate a security group, which is serialized and deserialized together with the ALB itself.
+
+```ts
+class ApplicationLoadBalancer {
+  serialize(ctx: ISerializationContext): void {
+    ctx.writeString('LoadBalancerArn', this.loadBalancerArn);
+    ctx.writeObject('SecurityGroup', this.securityGroup);
+  }
+
+  static deserializeApplicationLoadBalancer(ctx: IDeserializationContext): IApplicationLoadBalancer {
+    const exists = ctx.scope.findChild(ctx.id);
+    if (exists) {
+      return exists;
+    }
+
+    return new ImportedApplicationLoadBalancer(ctx.scope, ctx.id, {
+      loadBalancerArn: ctx.readString('LoadBalancerArn'),
+      securityGroup: ec2.SecurityGroup.deserializeSecurityGroup(ctx.readObject('SecurityGroup'))
+    });
+  }
+}
+```
+
+### Imports/Exports
+
+Now that constructs can be serialized and deserialized into a key-value context, we can implement a serialization mechanism for exporting resources from stacks and importing them in another stack. Importing can be done either at synthesis time using an environmental context provider or at runtime by returning `Fn::ImportValue` tokens for `readString`.
+
+The following methods will be added to the `Stack` class:
+
+```ts
+class Stack {
+  exportString(exportName: string, value: string, options?: ExportOptions): void;
```

Comment:
Why is this limited to `String`s? Why not have similar things for `Object`s? Why not `Stack.serializationContext(id: string)` and  `Stack.deserializationContext(id: string)` or something?